File: src/main/java/org/azamov/learnjakarta/task7_1/services/GroupService.java
package org.azamov.learnjakarta.task7_1.services;

import jakarta.persistence.EntityManager;
import org.azamov.learnjakarta.task7_1.helper.EntityManagerUtil;
import org.azamov.learnjakarta.task7_1.model.Group;
import org.azamov.learnjakarta.task7_1.model.Student;

import java.util.List;

public class GroupService {

    public void createGroup(String name, int createdBy) {
        try (EntityManager em = EntityManagerUtil.getEntityManager()) {
            em.getTransaction().begin();
            em.persist(Group.builder().name(name).createdBy(createdBy).build());
            em.getTransaction().commit();
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    public void deleteGroupById(int id) {
        try (EntityManager em = EntityManagerUtil.getEntityManager()) {
            em.getTransaction().begin();
            em.createQuery("DELETE FROM Groups g WHERE g.id = :id")
                    .setParameter("id", id)
                    .executeUpdate();
            em.getTransaction().commit();
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    public List<Student> getStudentsByGroupId(int groupId) {
        if (groupId <= 0) throw new IllegalArgumentException("groupId must be positive");
        try (EntityManager em = EntityManagerUtil.getEntityManager()) {
            return em.createQuery("SELECT s FROM Student s WHERE s.groupId = :groupId", Student.class)
                    .setParameter("groupId", groupId)
                    .getResultList();
        }
    }

    public List<Group> getGroups() {
        try (EntityManager em = EntityManagerUtil.getEntityManager()) {
            return em.createQuery("SELECT g FROM Groups g", Group.class)
                    .getResultList();
        }
    }
}

File: src/main/java/org/azamov/learnjakarta/task3/services/GroupService.kt
package org.azamov.learnjakarta.task3.services

import org.azamov.learnjakarta.task3.helper.toGroup
import org.azamov.learnjakarta.task3.helper.toStudent
import org.azamov.learnjakarta.task3.model.Group
import org.azamov.learnjakarta.task3.model.Student
import org.postgresql.Driver
import java.sql.Connection
import java.sql.DriverManager

class GroupService {
    init {
        DriverManager.registerDriver(Driver())
    }

    @Throws(Exception::class)
    private fun getConnection(): Connection {
        return DriverManager.getConnection(
            "jdbc:postgresql://localhost:5432/jakarta?currentSchema=jdbc_example", "postgres", "2255"
        )
    }

    fun createGroup(name: String, createdBy: Int): Boolean {
        getConnection().use { connection ->
            val stmt = connection.prepareStatement("INSERT INTO groups(name, created_by) VALUES (?, ?)")
            stmt.setString(1, name)
            stmt.setInt(2, createdBy)
            return stmt.executeUpdate() > 0
        }
    }

    fun deleteGroupById(id: Int): Int {
        getConnection().use { connection ->
            val stmt = connection.prepareStatement("DELETE FROM groups WHERE id = ?")
            stmt.setInt(1, id)
            return stmt.executeUpdate()
        }
    }

    fun getStudentsByGroupId(groupId: Int): List<Student> {
        require(groupId > 0) { "groupId must be positive" }
        val connection = getConnection()
        val statement = connection.createStatement()
        val students = ArrayList<Student>()
        val resultSet = statement.executeQuery("SELECT * FROM students where group_id = '$groupId';")
        while (resultSet.next()) {
            students.add(
                resultSet.toStudent()
            )
        }
        return students
    }

    fun getGroups(): List<Group> {
        val connection = getConnection()
        val statement = connection.createStatement()
        val resultSet = statement.executeQuery("SELECT * FROM groups")
        val groups = ArrayList<Group>()
        while (resultSet.next()) {
            groups.add(
                resultSet.toGroup()
            )
        }
        return groups
    }
}